### PR TITLE
Fix warnings from non-debug build with new Apple-clang

### PR DESCRIPTION
**Pull request by v-dobrev**
Mostly, these are "variable set but not used" warnings.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3578](https://github.com/mfem/mfem/pull/3578) | @v-dobrev | @pazner | @pazner + @jandrej | 4/5/23 | 4/5/23 | 4/10/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by pazner** [comment]
@v-dobrev, which version of clang are you using and what are the warning flags? I would like to try to reproduce locally.

**Comment by v-dobrev** [comment]
I use the flags `-pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter`. For some of these warnings one has to build with `MFEM_DEBUG=NO` to see them.

Regarding the compiler: now I have `Apple clang version 14.0.3 (clang-1403.0.22.14.1)` but I think the previous version `14.0.0` was showing the same warnings.

**Comment by pazner** [comment]
Hmm, I can't reproduce with `clang version 14.0.0 (clang-1400.0.29.202)` and `-pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter`

**Comment by v-dobrev** [comment]
Here's what I get on `master` with `MFEM_DEBUG=NO` building just the library (no examples/miniapps/tests):
```console
...
mpicxx  -pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter -O3 -std=c++11 -I./../hypre/src/hypre/include  -g  -c mesh/mesh.cpp -o mesh/mesh.o
mesh/mesh.cpp:5470:25: warning: variable 'fo' set but not used [-Wunused-but-set-variable]
   int i, j, k, wo = 0, fo = 0;
                        ^
1 warning generated.
...
mpicxx  -pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter -O3 -std=c++11 -I./../hypre/src/hypre/include  -g  -c mesh/mesh_readers.cpp -o mesh/mesh_readers.o
mesh/mesh_readers.cpp:1605:14: warning: variable 'elem_domain' set but not used [-Wunused-but-set-variable]
         int elem_domain; // another element's attribute (rarely used)
             ^
mesh/mesh_readers.cpp:1606:14: warning: variable 'n_partitions' set but not used [-Wunused-but-set-variable]
         int n_partitions; // number of partitions where an element takes place
             ^
2 warnings generated.
mpicxx  -pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter -O3 -std=c++11 -I./../hypre/src/hypre/include  -g  -c mesh/ncmesh.cpp -o mesh/ncmesh.o
mesh/ncmesh.cpp:6182:8: warning: variable 'count' set but not used [-Wunused-but-set-variable]
   int count = 0;
       ^
1 warning generated.
...
mpicxx  -pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter -O3 -std=c++11 -I./../hypre/src/hypre/include  -g  -c mesh/pncmesh.cpp -o mesh/pncmesh.o
mesh/pncmesh.cpp:1837:8: warning: variable 'nsent' set but not used [-Wunused-but-set-variable]
   int nsent = 0, nrecv = 0; // for debug check
       ^
mesh/pncmesh.cpp:1837:19: warning: variable 'nrecv' set but not used [-Wunused-but-set-variable]
   int nsent = 0, nrecv = 0; // for debug check
                  ^
2 warnings generated.
...
mpicxx  -pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter -O3 -std=c++11 -I./../hypre/src/hypre/include  -g  -c fem/tmop_tools.cpp -o fem/tmop_tools.o
fem/tmop_tools.cpp:158:13: warning: variable 'ti' set but not used [-Wunused-but-set-variable]
   for (int ti = 1; !last_step; ti++)
            ^
1 warning generated.
...
```
After that building all examples/miniapps/tests with `make all` gives:
```console
...
mpicxx -pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter -O3 -std=c++11 -I../.. -I../../../hypre/src/hypre/include  -g -c rubik.cpp
rubik.cpp:1109:8: warning: variable 'count' set but not used [-Wunused-but-set-variable]
   int count = 0;
       ^
1 warning generated.
...
mpicxx -g -pedantic -Wall -Wextra -Wshadow -Wno-unused-parameter -O3 -std=c++11 -I../.. -I../../../hypre/src/hypre/include  -g -I. -I../.. -c /<truncated>/mfem/tests/unit/fem/test_project_bdr.cpp -o fem/test_project_bdr.o
/<truncated>/mfem/tests/unit/fem/test_project_bdr.cpp:35:8: warning: variable 'npts' set but not used [-Wunused-but-set-variable]
   int npts = 0;
       ^
1 warning generated.
...
```
(I shortened some paths with `<truncated>`.)

File: fem/tmop_tools.cpp [hunk 1]
```diff
@@ -155,7 +155,7 @@ void AdvectorCG::ComputeAtNewPositionScalar(const Vector &new_nodes,
 
    double t = 0.0;
    bool last_step = false;
-   for (int ti = 1; !last_step; ti++)
+   for ( ; !last_step; )
```
Comment (pazner) [inline fem/tmop_tools.cpp:158]:
```suggestion
   while (!last_step)
```

Equivalent, but more idiomatic?

**Review by pazner — APPROVED** [review]
Some of these warnings must be new in clang 14.0.3, but changes all look good to me.

**Review by jandrej — APPROVED** [review]

File: fem/tmop_tools.cpp (hunk 1 above)
Comment (jandrej) [reply fem/tmop_tools.cpp:158]: I'm fine with it either way.

**Comment by v-dobrev** [comment]
> Some of these warnings must be new in clang 14.0.3, but changes all look good to me.

According to https://en.wikipedia.org/wiki/Xcode, it looks like Xcode 14.3 (clang 14.0.3) switched from LLVM 14 to 15 -- that's probably the main reason for these differences.

**Comment by pazner** [comment]
Merged in `next` for testing...

**Pull request merged**